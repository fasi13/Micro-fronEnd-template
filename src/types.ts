export interface Hierarchy {
	name: string;
}

export interface MfeOption {
	description: string;
	logo: string;
	order: number;
}

export interface Link {
	href: string;
	method: { method: 'POST' | 'GET' | 'PUT' | 'DELETE' };
	name: string;
	rel: string;
}

export interface ApiResponse<T> {
	data: T;
	success: boolean;
}

export interface DataPaginated<T> {
	items: T[];
	limit: number;
	offset: number;
	totalCount: number;
	_links: Link[];
}

export interface Application {
	id: number | string;
	name: string;
	key?: string;
	value: string;
	_links: Link[];
}
export interface TreeView {
	id: string | number;
	key?: string;
	name: string;
	value?: string;
	collapsed?: boolean;
	executedFetch?: boolean;
	isGroup?: boolean;
	childrenData?: TreeView[];
	parentId?: string | number;
	loading?: boolean;
	selfLink?: Link;
	selfUpdateLink?: Link;
	createGroupLink?: Link;
	createApplicationLink?: Link;
	loadChildrenLink?: Link[];
	path?: Application; // ??
	_links?: Link[];
	edit: boolean;
}

export interface ApplicationResponse {
	_links: Link[];
	id: number;
	key: string;
	name: string;
	value: string;
	parentApplicationId: number | null;
	applicationGroupId: number | null;
	dateModified: string;
}

export interface Wee {
	name: string;
}

export interface ItreeData {
	name: string;
	key: number;
	children: ItreeData[];
	edit: boolean;
}

// export interface ApplicationGroupResponse {
// 	_links: Link[];
// 	id: number;
// 	key: string;
// 	name: string;
// 	value: string;
// 	parentApplicationId: number | null;
// 	applicationGroupId: number | null;
// 	dateModified: string;
// }

// export interface IApplication {}
// export interface Group {}

// export interface Node {}
