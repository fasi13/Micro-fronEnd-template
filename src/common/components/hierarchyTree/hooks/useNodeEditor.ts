import { useRef, useState } from 'react';
import { ErrorResponse } from '../../../../types';

export interface useNodeEditorProps {
	data: string | undefined;
	isApplication: boolean;
	setError: (val: ErrorResponse | string) => void;
	clearError: () => void;
	onSubmit: (value: string) => void;
}

interface useNodeEditorReturnType {
	checkValidityAndSubmit: () => void;
	preValue: React.MutableRefObject<string>;
	value: string;
	nodeEditorPlaceHolder: () => string;
	setValue: React.Dispatch<React.SetStateAction<string>>;
}

export const useNodeEditor = (
	props: useNodeEditorProps,
): useNodeEditorReturnType => {
	const { data, isApplication, setError, clearError, onSubmit } = props;

	const [value, setValue] = useState(data || '');
	const preValue = useRef(value);

	const validAppPattern = /^[^)(]+(\([^()]+\)){1}$/g;
	const emptyErrorMsg = 'Value can not be empty';
	const invalidApplicationValueFormatMsg =
		'Application format should be: Application Name (Value)';

	const checkValidityAndSubmit = (): void => {
		if (value === '') {
			setError(emptyErrorMsg);
		} else if (
			isApplication &&
			value.match(validAppPattern)?.filter(m => m !== undefined).length !== 1
		) {
			setError(invalidApplicationValueFormatMsg);
		} else if (preValue.current !== value) {
			clearError();
			preValue.current = value;
			onSubmit(value);
		}
	};

	const nodeEditorPlaceHolder = (): string => {
		if (isApplication) return 'Add New Application';
		return 'Add New Application Group';
	};

	return {
		checkValidityAndSubmit,
		nodeEditorPlaceHolder,
		preValue,
		value,
		setValue,
	};
};
