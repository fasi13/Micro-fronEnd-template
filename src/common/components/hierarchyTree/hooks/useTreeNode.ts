// import { useCallback, useEffect, useReducer } from 'react';
// import { useBreadcrumbStore, useHierarchyStore } from '../../../../state';
// import {
// 	ErrorResponse,
// 	NodeActions,
// 	NodePath,
// 	TreeView,
// } from '../../../../types';

// type TEditor = '' | 'Application' | 'Group';
// type ActionType =
// 	| { type: 'LOAD_CHILDREN' }
// 	| { type: 'LOAD_CHILDREN_SUCCESS' }
// 	| { type: 'LOAD_CHILDREN_ERROR'; val: ErrorResponse }
// 	| { type: 'TOGGLE_NEW_EDITOR'; val: TEditor }
// 	| { type: 'TOGGLE_EDIT'; val: boolean }
// 	| { type: 'SAVE'; val: string }
// 	| { type: 'SAVING_SUCCESS' }
// 	| { type: 'SAVING_ERROR'; val: ErrorResponse }
// 	| { type: 'SAVE_RESET' }
// 	| { type: 'SET_ERROR'; val: ErrorResponse | string }
// 	| { type: 'CLEAR_ERROR' }
// 	| { type: 'COLLAPSE_CHILDREN_NODES' };

// interface TNodeDetail extends NodeActions {
// 	data: TreeView;
// 	nodeId: number;
// 	nodePath: NodePath[];
// 	edit: boolean | null;
// 	saving: boolean;
// 	toggleNewEditor: TEditor;
// 	newValue: string | null;
// 	error: string | null;
// 	loadingChildren: boolean;
// 	isFetched: boolean;
// 	nodeType: 'Application' | 'Group';
// }
// export interface TResponse {
// 	nodeState: TNodeDetail;
// 	toggleNewEditor: (val: TEditor) => void;
// 	editNode: (val: boolean) => void;
// 	toggleChildren: () => void;
// 	save: (value: string) => void;
// 	setError: (val: ErrorResponse | string) => void;
// 	clearError: () => void;
// }

// const extractApplicationNameAndValue = (text: string): string[] => {
// 	const newLocal = /^[^)(]+|\([^)(]*\)$/g;
// 	return text.match(new RegExp(newLocal))?.map(x => x.toString()) || ['', ''];
// };

// const reducer = (prevState: TNodeDetail, action: ActionType): TNodeDetail => {
// 	switch (action.type) {
// 		case 'LOAD_CHILDREN': {
// 			return {
// 				...prevState,
// 				loadingChildren: true,
// 				toggleNewEditor: '',
// 			};
// 		}
// 		case 'LOAD_CHILDREN_SUCCESS': {
// 			return { ...prevState, loadingChildren: false, isFetched: true };
// 		}
// 		case 'LOAD_CHILDREN_ERROR': {
// 			return {
// 				...prevState,
// 				loadingChildren: false,
// 				error: action.val.errors?.[0],
// 				isFetched: false,
// 			};
// 		}
// 		case 'TOGGLE_EDIT':
// 			return { ...prevState, edit: action.val, toggleNewEditor: '' };
// 		case 'TOGGLE_NEW_EDITOR': {
// 			if (action.val) return { ...prevState, toggleNewEditor: action.val };

// 			return {
// 				...prevState,
// 				toggleNewEditor: action.val,
// 				error: null,
// 			};
// 		}
// 		case 'SAVE': {
// 			return { ...prevState, newValue: action.val, saving: true, error: null };
// 		}
// 		case 'SAVING_SUCCESS':
// 			return {
// 				...prevState,
// 				saving: false,
// 				toggleNewEditor: '',
// 				loadingChildren: false,
// 				error: null,
// 				isFetched: false,
// 			};
// 		case 'SAVING_ERROR':
// 			return { ...prevState, saving: false, error: action.val.errors?.[0] };
// 		case 'SAVE_RESET':
// 			return {
// 				...prevState,
// 				saving: false,
// 				toggleNewEditor: '',
// 				loadingChildren: false,
// 				error: null,
// 			};
// 		case 'SET_ERROR': {
// 			if (typeof action.val === 'string')
// 				return { ...prevState, error: action.val };
// 			return { ...prevState, error: action.val.errors?.[0] };
// 		}
// 		case 'CLEAR_ERROR':
// 			return { ...prevState, error: null };
// 		default:
// 			throw new Error();
// 	}
// };

// export const useTreeNode = (nodeDetail: TNodeDetail): TResponse => {
// 	const [nodeState, dispatch] = useReducer(reducer, nodeDetail);
// 	const { setBreadCrumb } = useBreadcrumbStore();

// 	const toggleNewEditor = useCallback((val: TEditor) => {
// 		dispatch({ type: 'TOGGLE_NEW_EDITOR', val });
// 	}, []);

// 	const editNode = useCallback((val: boolean) => {
// 		dispatch({ type: 'TOGGLE_EDIT', val });
// 	}, []);

// 	const save = useCallback((value: string) => {
// 		dispatch({ type: 'SAVE', val: value });
// 	}, []);

// 	const setError = useCallback((val: ErrorResponse | string) => {
// 		dispatch({ type: 'SET_ERROR', val });
// 	}, []);

// 	const clearError = useCallback(() => {
// 		dispatch({ type: 'CLEAR_ERROR' });
// 	}, []);

// 	const savingCb = useCallback((err: ErrorResponse | null) => {
// 		if (err) {
// 			dispatch({ type: 'SAVING_ERROR', val: err });
// 		} else if (nodeState.edit === false) {
// 			// refetch
// 			dispatch({ type: 'SAVING_SUCCESS' });
// 			// call  state method to collapse children
// 			// store.collapseChildren(node);
// 		}
// 	}, []);

// 	const loadNodeChildren = useCallback(() => {
// 		const { data, nodeId, nodePath, getChildren, onToggleCollapse } =
// 			nodeDetail;

// 		getChildren(data, nodeId, nodePath, (err: ErrorResponse | null) => {
// 			if (err) dispatch({ type: 'LOAD_CHILDREN_ERROR', val: err });
// 			else {
// 				dispatch({ type: 'LOAD_CHILDREN_SUCCESS' });
// 				onToggleCollapse(nodePath, false);
// 				setBreadCrumb(nodePath);
// 			}
// 		});
// 	}, [nodeState.loadingChildren]);

// 	const editApplication = useCallback(
// 		(name: string, value: string) => {
// 			if (nodeState.saving) {
// 				const { data, nodeId, nodePath, onEditApplication } = nodeDetail;
// 				onEditApplication(
// 					data,
// 					nodeId,
// 					nodePath,
// 					name,
// 					value.replaceAll(/\(|\)/gi, ''),
// 					savingCb,
// 				);
// 			}
// 		},
// 		[nodeState.saving],
// 	);

// 	const editGroup = useCallback(
// 		(name: string) => {
// 			if (nodeState.saving) {
// 				const { data, nodeId, nodePath, onEditGroup } = nodeDetail;
// 				onEditGroup(data, nodeId, nodePath, name, savingCb);
// 			}
// 		},
// 		[nodeState.saving],
// 	);

// 	const addApplication = useCallback(
// 		(name: string, value: string) => {
// 			if (nodeState.saving) {
// 				const { data, nodeId, nodePath, onAddApplication } = nodeDetail;
// 				onAddApplication(
// 					data,
// 					nodeId,
// 					nodePath,
// 					name,
// 					value.replaceAll(/\(|\)/gi, ''),
// 					savingCb,
// 				);
// 			}
// 		},
// 		[nodeState.saving],
// 	);

// 	const addGroup = useCallback(
// 		(name: string) => {
// 			if (nodeState.saving) {
// 				const { data, nodeId, nodePath, onAddGroup } = nodeDetail;
// 				onAddGroup(data, nodeId, nodePath, name, savingCb);
// 			}
// 		},
// 		[nodeState.saving],
// 	);

// 	const toggleChildren = useCallback(() => {
// 		const { data, nodePath, onToggleCollapse } = nodeDetail;

// 		if (data.collapsed && !nodeState.isFetched && !data.childrenData)
// 			dispatch({ type: 'LOAD_CHILDREN' });
// 		else if (data.collapsed && nodeState.isFetched && data.childrenData)
// 			onToggleCollapse(nodePath, false);
// 		else if (!data.collapsed) {
// 			// just collapse
// 			onToggleCollapse(nodePath, true);
// 		}
// 	}, [nodeDetail]);

// 	useEffect(() => {
// 		if (nodeState.saving) {
// 			const { newValue } = nodeState;
// 			if (nodeState.edit) {
// 				if (nodeState.nodeType === 'Application' && newValue) {
// 					const [name, value] = extractApplicationNameAndValue(newValue);
// 					editApplication(name, value);
// 				} else if (nodeState.nodeType === 'Group' && newValue) {
// 					editGroup(newValue);
// 				}
// 			} else if (nodeState.toggleNewEditor === 'Application' && newValue) {
// 				const [name, value] = extractApplicationNameAndValue(newValue);
// 				addApplication(name, value);
// 			} else if (nodeState.toggleNewEditor === 'Group' && newValue) {
// 				addGroup(newValue);
// 			}
// 		}
// 	}, [nodeState.saving]);

// 	useEffect(() => {
// 		if (nodeState.loadingChildren) loadNodeChildren();
// 	}, [nodeState.loadingChildren]);

// 	useEffect(() => {
// 		if (!nodeState.edit) dispatch({ type: 'CLEAR_ERROR' });
// 	}, [nodeState.edit]);

// 	useEffect(() => {
// 		console.log('???? -->', nodeDetail.data);
// 	}, []);

// 	return {
// 		nodeState,
// 		toggleNewEditor,
// 		toggleChildren,
// 		editNode,
// 		save,
// 		setError,
// 		clearError,
// 	};
// };


export { };
