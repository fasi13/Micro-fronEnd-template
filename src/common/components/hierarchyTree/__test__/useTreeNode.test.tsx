import { renderHook } from '@testing-library/react-hooks';
import { TreeView } from '../../../../types';
import { TUseTreeNodeProps, useTreeNode } from '../hooks/useTreeNode';

let dummyProps: TUseTreeNodeProps;
let dummyTreeView: TreeView;
const selfUpdateStr = 'dummy/selfupdate';
const createGroupStr = 'dummy/createGroup';
const createAppStr = 'dummy/createApp';

describe('useTreeNode', () => {
	beforeEach(() => {
		jest.clearAllMocks();
		dummyTreeView = {
			id: 1,
			name: 'dummy-name',
			value: 'dummy-value',
			collapsed: false,
			edit: false,
			error: null,
			saving: false,
			loadingChildren: false,
			toggleNewEditor: '',
			nodeDepth: 0,
			nodePath: [{ pathId: 1, pathName: 'dummy' }],
			_links: [
				{
					href: selfUpdateStr,
					method: { method: 'GET' },
					name: 'updateSelf',
					rel: 'updateApplicationGroup',
				},
				{
					href: selfUpdateStr,
					method: { method: 'GET' },
					name: 'updateSelf',
					rel: 'updateApplication',
				},

				{
					href: createGroupStr,
					method: { method: 'POST' },
					name: 'createGroup',
					rel: 'createApplicationGroup',
				},
				{
					href: createGroupStr,
					method: { method: 'POST' },
					name: 'createGroup',
					rel: '',
				},
				{
					href: createAppStr,
					method: { method: 'POST' },
					name: 'createApplication',
					rel: 'createApplication',
				},
				{
					href: 'dummy/children_application',
					method: { method: 'GET' },
					name: 'getChildren',
					rel: 'applications',
				},
				{
					href: 'dummy/children_applicationGroup',
					method: { method: 'GET' },
					name: 'getChildren',
					rel: 'applicationGroups',
				},
			],
		};
		dummyProps = {
			data: { ...dummyTreeView },
			nodePath: dummyTreeView.nodePath,
			onToggleCollapse: jest.fn(),
			onToggleNewEditor: jest.fn(),
			onToggleEdit: jest.fn(),
			onSetSaving: jest.fn(),
			onSetNodeErr: jest.fn(),
			onAddApplication: jest.fn(),
			onEditApplication: jest.fn(),
			onAddGroup: jest.fn(),
			onEditGroup: jest.fn(),
		};
	});

	describe('useTreeNode snapShot return values', () => {
		it('returns object of helper methods', () => {
			const result = renderHook(useTreeNode, { initialProps: dummyProps });
			expect(result).toMatchSnapshot();
		});
	});

	describe('canAddApplication', () => {
		it('returns true if node has link with rel createApplication', () => {
			const { result } = renderHook(useTreeNode, {
				initialProps: dummyProps,
			});

			expect(result.current.canAddApplication(dummyTreeView)).toBe(true);
		});
	});

	describe('showHideNewEditorAndTreeChildren', () => {
		it('returns mb-10 if is collapse is false and editorMode is not empty', () => {
			const { result } = renderHook(useTreeNode, {
				initialProps: dummyProps,
			});
			expect(
				result.current.showHideNewEditorAndTreeChildren(false, 'Application'),
			).toBe('mb-10');
		});

		it('returns empty string if is collapsed is false and editorMode is empty', () => {
			const { result } = renderHook(useTreeNode, {
				initialProps: dummyProps,
			});
			expect(result.current.showHideNewEditorAndTreeChildren(false, '')).toBe(
				'',
			);
		});

		it('returns hidden text if is collapsed is true', () => {
			const { result } = renderHook(useTreeNode, {
				initialProps: dummyProps,
			});
			expect(result.current.showHideNewEditorAndTreeChildren(true, '')).toBe(
				'hidden',
			);
		});
	});

	describe('nodeValue', () => {
		it('returns name as it is if it is group', () => {
			const { result } = renderHook(useTreeNode, {
				initialProps: dummyProps,
			});
			expect(result.current.nodeValue()).toEqual('dummy-name');
		});

		it('returns name with value if it is application', () => {
			dummyProps.data._links = dummyProps.data?._links.filter(
				l => l.rel !== 'createApplication',
			);
			const { result } = renderHook(useTreeNode, {
				initialProps: dummyProps,
			});
			expect(result.current.nodeValue()).toEqual('dummy-name (dummy-value)');
		});
	});

	describe('extractApplicationNameAndValue', () => {
		it(`given 'name (value)' it returns array of ['name','value']`, () => {
			const { result } = renderHook(useTreeNode, {
				initialProps: dummyProps,
			});

			const val: string[] =
				result.current.extractApplicationNameAndValue('name (value)');

			expect(val).toContain('name ');
			expect(val).toContain('value');
		});

		it(`given 'dummytext' it returns array of ['dummytext']`, () => {
			const { result } = renderHook(useTreeNode, {
				initialProps: dummyProps,
			});

			const val: string[] =
				result.current.extractApplicationNameAndValue('dummytext');

			expect(val.length).toBe(2);
			expect(val[0]).toContain('dummytext');
			expect(val[1]).toContain('dummytext');

		});

		it(`given '' it returns array of ['','']`, () => {
			const { result } = renderHook(useTreeNode, {
				initialProps: dummyProps,
			});

			const val: string[] = result.current.extractApplicationNameAndValue('');

			expect(val).toContain('');
			expect(val).toContain('');
		});
	});

	describe('submitHandler', () => {
		it('calls onEditGroup', () => {
			const { result } = renderHook(useTreeNode, {
				initialProps: { ...dummyProps, data: { ...dummyTreeView, edit: true } },
			});

			result.current.submitHandler('newValue');

			expect(dummyProps.onEditGroup).toHaveBeenCalled();
		});

		it('calls onEditApplication', () => {
			const { result } = renderHook(useTreeNode, {
				initialProps: {
					...dummyProps,
					data: {
						...dummyTreeView,
						edit: true,
						_links: dummyTreeView?._links.filter(
							l => l.rel !== 'createApplication',
						),
					},
				},
			});

			result.current.submitHandler('newValue');

			expect(dummyProps.onEditApplication).toHaveBeenCalled();
		});

		it('calls onAddApplication', () => {
			const { result } = renderHook(useTreeNode, {
				initialProps: {
					...dummyProps,
					data: {
						...dummyTreeView,
						edit: false,
						toggleNewEditor: 'Application',
					},
				},
			});

			result.current.submitHandler('newValue');

			expect(dummyProps.onAddApplication).toHaveBeenCalled();
		});

		it('calls onAddGroup', () => {
			const { result } = renderHook(useTreeNode, {
				initialProps: {
					...dummyProps,
					data: {
						...dummyTreeView,
						edit: false,
						toggleNewEditor: 'Group',
					},
				},
			});

			result.current.submitHandler('newValue');

			expect(dummyProps.onAddGroup).toHaveBeenCalled();
		});

		it('calls nothing if it is not in edit mode and toggleNewEditor emptyString', () => {
			const { result } = renderHook(useTreeNode, {
				initialProps: {
					...dummyProps,
					data: {
						...dummyTreeView,
						edit: false,
						toggleNewEditor: '',
					},
				},
			});

			result.current.submitHandler('newValue');

			expect(dummyProps.onEditGroup).not.toHaveBeenCalled();
			expect(dummyProps.onEditApplication).not.toHaveBeenCalled();
			expect(dummyProps.onAddApplication).not.toHaveBeenCalled();
			expect(dummyProps.onAddGroup).not.toHaveBeenCalled();
		});
	});

	describe('closeEditor', () => {
		it('calls onToggleEdit', () => {
			const { result } = renderHook(useTreeNode, {
				initialProps: dummyProps,
			});

			result.current.closeEditor();

			expect(dummyProps.onToggleEdit).toHaveBeenCalled();
		});
	});

	describe('setErrorHandler', () => {
		it('calls onSetNodeErr', () => {
			const { result } = renderHook(useTreeNode, {
				initialProps: dummyProps,
			});

			result.current.setErrorHandler('dummy-error');

			expect(dummyProps.onSetNodeErr).toHaveBeenCalledWith(
				[{ pathId: 1, pathName: 'dummy' }],
				'dummy-error',
			);
		});
	});

	describe('clearErrorHandler', () => {
		it('calls onSetNodeErr', () => {
			const { result } = renderHook(useTreeNode, {
				initialProps: dummyProps,
			});

			result.current.clearErrorHandler();

			expect(dummyProps.onSetNodeErr).toHaveBeenCalled();
		});
	});

	describe('toggleEdit', () => {
		it('toggleEdit calls onToggleEdit', () => {
			const { result } = renderHook(useTreeNode, {
				initialProps: dummyProps,
			});
			const spy = jest.spyOn(result.current, 'toggleEdit');
			result.current.toggleEdit();
			expect(spy).toHaveBeenCalled();
		});
	});

	describe('toggleChildren', () => {
		it('toggleChildren calls onToggleEdit', () => {
			const { result } = renderHook(useTreeNode, {
				initialProps: dummyProps,
			});
			const spy = jest.spyOn(result.current, 'toggleChildren');
			result.current.toggleChildren();
			expect(spy).toHaveBeenCalled();
		});
	});

	describe('toggleNewEditor', () => {
		it('toggleNewEditor calls onToggleEdit', () => {
			const { result } = renderHook(useTreeNode, {
				initialProps: dummyProps,
			});
			const spy = jest.spyOn(result.current, 'toggleNewEditor');
			result.current.toggleNewEditor('Application');
			expect(spy).toHaveBeenCalled();
		});
	});

	describe('closeNewEditor', () => {
		it('closeNewEditor calls onToggleEdit', () => {
			const { result } = renderHook(useTreeNode, {
				initialProps: dummyProps,
			});
			const spy = jest.spyOn(result.current, 'closeNewEditor');
			result.current.closeNewEditor();
			expect(spy).toHaveBeenCalled();
		});
	});
});
