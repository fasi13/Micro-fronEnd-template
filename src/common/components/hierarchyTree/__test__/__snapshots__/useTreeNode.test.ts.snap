// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTreeNode useTreeNode snapShot return values returns object of helper methods 1`] = `
Object {
  "rerender": [Function],
  "result": Object {
    "all": Array [
      Object {
        "canAddApplication": [Function],
        "clearErrorHandler": [Function],
        "closeEditor": [Function],
        "closeNewEditor": [Function],
        "extractApplicationNameAndValue": [Function],
        "isApplication": [Function],
        "nodeValue": [Function],
        "setErrorHandler": [Function],
        "showHideNewEditorAndTreeChildren": [Function],
        "submitHandler": [Function],
        "toggleChildren": [Function],
        "toggleEdit": [Function],
        "toggleNewEditor": [Function],
      },
    ],
    "current": Object {
      "canAddApplication": [Function],
      "clearErrorHandler": [Function],
      "closeEditor": [Function],
      "closeNewEditor": [Function],
      "extractApplicationNameAndValue": [Function],
      "isApplication": [Function],
      "nodeValue": [Function],
      "setErrorHandler": [Function],
      "showHideNewEditorAndTreeChildren": [Function],
      "submitHandler": [Function],
      "toggleChildren": [Function],
      "toggleEdit": [Function],
      "toggleNewEditor": [Function],
    },
    "error": undefined,
  },
  "unmount": [Function],
  "waitFor": [Function],
  "waitForNextUpdate": [Function],
  "waitForValueToChange": [Function],
}
`;
